compiler/aarch64/codegen/OMRMachine.cpp

```diff
@@ -79,18 +82,380 @@ TR::RealRegister *OMR::ARM64::Machine::freeBestRegister(TR::Instruction *current
+            }
+         }
+      }
+
+   if (self()->cg()->comp()->getOption(TR_TraceCG))
+      {
+      diagnostic("\n\tspilling %s (%s)",
+                  best->getAssignedRegister()->getRegisterName(self()->cg()->comp()),
+                  best->getRegisterName(self()->cg()->comp()));
+      }
+
+   switch (rk)
+      {
+      case TR_GPR:
+         generateTrg1MemInstruction(self()->cg(), TR::InstOpCode::ldrimmx, currentNode, best, tmemref, currentInstruction);
+         break;
+      case TR_FPR:
+         generateTrg1MemInstruction(self()->cg(), TR::InstOpCode::vldrimmd, currentNode, best, tmemref, currentInstruction);
+         break;
+      default:
+         TR_ASSERT(false, "Unsupported RegisterKind.");
+         break;
+      }
+
+   self()->cg()->traceRegFreed(candidates[0], best);
+
+   best->setAssignedRegister(NULL);
+   best->setState(TR::RealRegister::Free);
+   candidates[0]->setAssignedRegister(NULL);
+   return best;
    }
 
 TR::RealRegister *OMR::ARM64::Machine::reverseSpillState(TR::Instruction *currentInstruction,
                                                      TR::Register *spilledRegister,
                                                      TR::RealRegister *targetRegister)
    {
-   TR_ASSERT(false, "Not implemented yet.");
+   TR::Compilation *comp = self()->cg()->comp();
+   TR::MemoryReference *tmemref;
+   TR::RealRegister *sameReg;
+   TR_BackingStore *location = spilledRegister->getBackingStorage();
+   TR::Node *currentNode = currentInstruction->getNode();
+   TR_RegisterKinds rk = spilledRegister->getKind();
+   TR_Debug *debugObj = self()->cg()->getDebug();
+
+   if (targetRegister == NULL)
+      {
+      targetRegister = self()->findBestFreeRegister(rk);
+      if (targetRegister == NULL)
+         {
+         targetRegister = self()->freeBestRegister(currentInstruction, spilledRegister, NULL);
+         }
+      targetRegister->setState(TR::RealRegister::Assigned);
+      }
+
+   if (self()->cg()->isOutOfLineColdPath())
+      {
+      // the future and total use count might not always reflect register spill state
+      // for example a new register assignment in the hot path would cause FC != TC
+      // in this case, assign a new register and return
+      if (!location)
+         {
+         if (debugObj)
+            self()->cg()->traceRegisterAssignment("OOL: Not generating reverse spill for (%s)\n", debugObj->getName(spilledRegister));
+         return targetRegister;
+         }
+      }
 
-   return NULL;
+   if (comp->getOption(TR_TraceCG))
+      {
+      diagnostic("\n\tre-assigning spilled %s to %s",
+                  spilledRegister->getRegisterName(comp),
+                  targetRegister->getRegisterName(comp));
+      }
+
+   tmemref = new (self()->cg()->trHeapMemory()) TR::MemoryReference(currentNode, location->getSymbolReference(), 8, self()->cg());
+
+   if (comp->getOption(TR_DisableOOL))
+      {
+      switch (rk)
+         {
+         case TR_GPR:
```

Comment:
Same as earlier comments about commoning these cases up.  Only difference seems to be the calculation of the spill size and the opcode.

Reply:
Merged the GPR and FPR cases.